### PR TITLE
feat: armor transparency

### PR DESCRIPTION
**Pull request by YouSoPugly**

### CODE REVIEW

File: common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java [hunk 1]
```diff
@@ -30,4 +34,13 @@ public void onTranslucentCheck(LivingEntityRenderTranslucentCheckEvent e) {
             e.setTranslucence(playerGhostTranslucenceLevel);
         }
     }
+
+    @SubscribeEvent
+    public void onGhostArmorRender(PlayerGhostArmorRenderEvent e) {
+        if (!WynnUtils.onWorld()) return;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java]: Events are only posted if this condition is met, remove this

File: common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java [hunk 2]
```diff
@@ -20,6 +21,9 @@ public class PlayerGhostTransparencyFeature extends UserFeature {
     @Config
     public float playerGhostTranslucenceLevel = 0.75f;
 
+    @Config
+    public boolean transparentPlayerGhostArmor = true;
+
     @SubscribeEvent
     public void onTranslucentCheck(LivingEntityRenderTranslucentCheckEvent e) {
         if (!WynnUtils.onWorld()) return;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java]: this one is left behind from long ago

File: common/src/main/java/com/wynntils/mc/EventFactory.java [hunk 3]
```diff
@@ -10,64 +10,16 @@
...
 import com.wynntils.core.WynntilsMod;
-import com.wynntils.mc.event.AddEntityLookupEvent;
-import com.wynntils.mc.event.ArmSwingEvent;
-import com.wynntils.mc.event.BossHealthUpdateEvent;
-import com.wynntils.mc.event.ChatPacketReceivedEvent;
-import com.wynntils.mc.event.ChatSentEvent;
-import com.wynntils.mc.event.ChestMenuQuickMoveEvent;
-import com.wynntils.mc.event.ClientTickEvent;
-import com.wynntils.mc.event.CommandsPacketEvent;
+import com.wynntils.mc.event.*;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/EventFactory.java]: No wildcard imports please

File: common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    public PlayerGhostArmorRenderEvent(LivingEntity entity) {
+        this.entity = entity;
+        this.renderGhostArmor = true;
+    }
+
+    public LivingEntity getEntity() {
+        return entity;
+    }
+
+    public boolean shouldRenderGhostArmor() {
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java]: I still would rather have the event cancelable, like I mentioned in the last PR

**Review by kristofbolyai — CHANGES_REQUESTED** [review]
Looks better, some changes are still needed

File: common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java (hunk 1 above)
Comment (magicus) [reply common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java]: Actually, events are only posted if `onServer`. If we want to disable this functionality on the hub, this check is needed. (I don't know if this is the case here, though)

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java]: Doesn't matter here

File: common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(HumanoidArmorLayer.class)
+public abstract class HumanoidArmorLayerMixin<T extends LivingEntity> {
+
+    @Inject(method = "render(Lcom/mojang/blaze3d/vertex/PoseStack;Lnet/minecraft/client/renderer/MultiBufferSource;ILnet/minecraft/world/entity/LivingEntity;FFFFFF)V", at = @At("INVOKE"), cancellable = true)
+    private void render(PoseStack matrixStack, MultiBufferSource buffer, int packedLight, T livingEntity, float limbSwing, float limbSwingAmount, float partialTicks, float ageInTicks, float netHeadYaw, float headPitch, CallbackInfo ci) {
+        if (!(livingEntity instanceof Player))
+            return;
+        PlayerGhostArmorRenderEvent event = EventFactory.onArmorRender(livingEntity);
```
Comment (magicus) [inline common/src/main/java/com/wynntils/mc/mixin/HumanoidArmorLayerMixin.java]: Cast livingEntity to Player here, where you made the check, and let the event and all subsequent users have a Player instead.

File: common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java [hunk 6]
```diff
@@ -30,4 +34,13 @@ public void onTranslucentCheck(LivingEntityRenderTranslucentCheckEvent e) {
             e.setTranslucence(playerGhostTranslucenceLevel);
         }
     }
+
+    @SubscribeEvent
+    public void onGhostArmorRender(PlayerGhostArmorRenderEvent e) {
+        if (!WynnUtils.onWorld()) return;
+
+        if (WynnPlayerUtils.isPlayerGhost((Player) e.getEntity())) {
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/user/players/PlayerGhostTransparencyFeature.java]: This cast does not look good. When reading this code, the question "how do you know this is a Player" pops into your mind, and the gut reaction is to add an instance test first. But I continued through the code, and found the answer in a very different place. I made a comment there about how this needs to be fixed. And when that is done, this cast can be removed.

**Comment by YouSoPugly** [comment]
added the fixes asked of me, lmk if there's anything else

File: common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java (hunk 4 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java]: Are you intentionally skipping this? Do you need help with it? Or maybe I didn't explain it cleary

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/mc/event/PlayerGhostArmorRenderEvent.java]: @YouSoPugly ^

**Review by kristofbolyai — APPROVED** [review]
Thank you for your contribution and for your patience.

**Pull request merged**